compiler/ballerina-lang/src/main/java/io/ballerina/projects/Project.java

```diff
@@ -152,4 +154,13 @@ protected Project resetPackage(Project project) {
     public abstract Optional<Path> documentPath(DocumentId documentId);
 
     public abstract void save();
+
+    /**
+     * Maintains the BuildContext for a project which acts as a data holder for a project.
+     *
+     * @return BuildContext
+     */
+    public BuildContext buildContext() {
+        return this.buildContext;
```

Comment:
I have a small concern regarding this. Looking at the issue, the support is specific to the compiler plugins. So, will this method be called for each Project instance? If not, we can follow lazy loading and simply return the instance without holding the variable.

```suggestion
        return BuildContext.getInstance();
```

Also, this would mean modifying the `getInstance` method to create the instance on the first call.

Reply:
It's a valid concern @nipunayf 
It's only applicable for compiler plugins. So, it makes sense to do lazy loading.

Reply:
Addressed with 6078cd3540c224d73238762953ecd356e8fd1e09